### DISCUSSION TITLE
OpenAI Chat GPT and http_request Post

### DISCUSSION THREAD

**Discussion by BigDarkAkula** [supabase/supabase#13210] [General]
Hello, I am trying to pull down an API response from Chat GPT and I am having some issues with this code:

```
CREATE OR REPLACE FUNCTION public.openai(
    prompt text
)
RETURNS text
LANGUAGE plpgsql
AS $$
DECLARE
    api_key text := (SELECT api_key FROM api_info WHERE system='openAI' and type='general' LIMIT 1);
    url text := 'https://api.openai.com/v1/completions';
    key_header text := 'Bearer ' || api_key;
    data jsonb := json_build_object(
                    'model', 'text-davinci-003',
                    'prompt', prompt
                );

    response jsonb;
    records text; -- declare a variable to store the result
BEGIN
    SELECT CONTENT::jsonb INTO records FROM http((
      'POST',
      url,
      ARRAY[http_header('Authorization', key_header)],
      'Content-Type: application/json',
      data
    )::http_request);

    RETURN records;
END;
$$;
```

I am getting an error message that says, "{"error": {"code": null, "type": "invalid_request_error", "param": null, "message": "you must provide a model parameter"}}"

There is not a ton of documentation on this http_request type that I see except for this:
https://github.com/pramsey/pgsql-http

The documentation says the structure should be like this:

```
     Composite type "public.http_request"
    Column    |       Type        | Modifiers
--------------+-------------------+-----------
 method       | http_method       |
 uri               | character varying |
 headers      | http_header[]     |
 content_type | character varying |
 content      | character varying |

```
Chat GPT told me the structure should be like this (which I am guessing is incorrect)
```

SELECT [response] INTO [variable_name] FROM http((
  [http_method],
  [url],
  [http_headers],
  [http_body],
  [http_options]
)::http_request)
```

The function didn't like me putting the content type in the headers section and I got an error until I put it in the third parameter from the documentation so I am guessing Chat GPT is wrong.  You can see the last two parameters and I am not sure how I need to format the body either.

Can someone give me a hint on how to get a response from Chat GPT using this?

**Comment by BigDarkAkula** [comment]
I made this function in Python 3.7 that works in AWS Lambda to call chatGTP.  Is there any way to move this over and use it in Supabase?  I am not really sure if I can use the extra packages in Supabase.  I still cannot get the postgres function above to work.

**Comment by ricgagliardi** [comment]
I ran into the same problem with an almost identical script. I can kludge it by calling a third-party microservice, like Deno from within a Postgreql function that would in turn call OpenAI. But that would leave a security hole.

I hope a solution arrives before too long!
